simd/unit_tests/include/TestSIMD_MathOps.hpp

```diff
@@ -73,40 +74,70 @@ void host_check_math_op_one_loader(UnaryOp unary_op, std::size_t n,
                                     const DataType (&second_args)[n]) {
-  host_check_math_op_all_loaders<Abi>(plus(), n, first_args, second_args);
-  host_check_math_op_all_loaders<Abi>(minus(), n, first_args, second_args);
-  host_check_math_op_all_loaders<Abi>(multiplies(), n, first_args, second_args);
-  host_check_math_op_all_loaders<Abi>(absolutes(), n, first_args);
-
-  host_check_math_op_all_loaders<Abi>(floors(), n, first_args);
-  host_check_math_op_all_loaders<Abi>(ceils(), n, first_args);
-  host_check_math_op_all_loaders<Abi>(rounds(), n, first_args);
-  host_check_math_op_all_loaders<Abi>(truncates(), n, first_args);
+  const int integral_ulps = 0;
+  host_check_math_op_all_loaders_binary<Abi>(plus(), n, first_args, second_args,
+                                             integral_ulps);
+  host_check_math_op_all_loaders_binary<Abi>(minus(), n, first_args,
+                                             second_args, integral_ulps);
+  host_check_math_op_all_loaders_binary<Abi>(multiplies(), n, first_args,
+                                             second_args, integral_ulps);
+  host_check_math_op_all_loaders_unary<Abi>(absolutes(), n, first_args,
+                                            integral_ulps);
+
+  host_check_math_op_all_loaders_unary<Abi>(floors(), n, first_args,
+                                            integral_ulps);
+  host_check_math_op_all_loaders_unary<Abi>(ceils(), n, first_args,
+                                            integral_ulps);
+  host_check_math_op_all_loaders_unary<Abi>(rounds(), n, first_args,
+                                            integral_ulps);
+  host_check_math_op_all_loaders_unary<Abi>(truncates(), n, first_args,
+                                            integral_ulps);
 
   // TODO: Place fallback implementations for all simd integer types
   if constexpr (std::is_floating_point_v<DataType>) {
-    host_check_math_op_all_loaders<Abi>(divides(), n, first_args, second_args);
-
-#if defined(__INTEL_COMPILER) && \
+    const int floating_ulps = 4;
+    host_check_math_op_all_loaders_binary<Abi>(divides(), n, first_args,
+                                               second_args, floating_ulps);
+#if (defined(__INTEL_COMPILER) || defined(__INTEL_LLVM_COMPILER)) && \
     (defined(KOKKOS_ARCH_AVX2) || defined(KOKKOS_ARCH_AVX512XEON))
-    host_check_math_op_all_loaders<Abi>(cbrt_op(), n, first_args);
-    host_check_math_op_all_loaders<Abi>(exp_op(), n, first_args);
-    host_check_math_op_all_loaders<Abi>(log_op(), n, first_args);
+    host_check_math_op_all_loaders_unary<Abi>(cbrt_op(), n, first_args,
+                                              floating_ulps);
+    host_check_math_op_all_loaders_unary<Abi>(exp_op(), n, first_args,
+                                              floating_ulps);
+    DataType positive_args[n];
+    for (size_t i = 0; i < n; ++i) {
+      positive_args[i] = std::abs(first_args[i]) + 0.001;
+    }
```

Comment:
what was the issue here?  we were calling log with arguments <= 0 ?

Reply:
Looking more into this
`log(0)` should return negative infinity.  What was wrong with it?

Reply:
```
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
/app/kokkos/simd/unit_tests/include/SIMDTesting_Utilities.hpp:35: Failure
The difference between a and b is nan, which exceeds std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)), where
a evaluates to -inf,
b evaluates to -inf, and
std::numeric_limits<T>::epsilon() * ulps * std::max(std::fabs(a), std::fabs(b)) evaluates to inf.
```

Reply:
What do we conclude from this?  That the floating-point comparison needs to be revised?

Reply:
I would just not care about testing non-positive arguments here.